storage/reader.go

```diff
@@ -392,6 +574,72 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return n, err
 }
 
+func (r *Reader) readWithGRPC(p []byte) (int, error) {
+	// No stream to read from, either never initiliazed or Close was called.
+	// TODO: should an error be returned here? "reader has been closed"?
```

Comment:
Yeah we should return an error probably; EOF seems bad because it implies that we finished reading the object. Or if "never initialized" somehow we'd want to try to reopen the stream probably?

Reply:
I'm going to have it return an error. I don't want to reopen b.c part of the `Close` implementation here will set `r.stream` to `nil` in order to have it be GC'd. So if `r.stream == nil`' the user called `Close` and a subsequent `Read` should be an error.

However there is a possible race condition in the event there are multiple consumers of the Reader and one call to Read encounters an error on `Recv` and the stream is to be reopened. If another consumer calls Read on the same Reader, the stream will either be `nil` or not functional/closed (by the backend). Is that scenario a concern @tritone ?

Reply:
Ah yeah that makes sense. As for the race condition, I think we don't make any thread safety guarantees about the current Reader implementation either -- it would be really weird to have multiple go routines reading from the same reader actually. Maybe we should note this explicitly in the docs.

Reply:
Adding a note about this.